### ISSUE TITLE
Completing a message from a streaming receiver with autoComplete enabled results in program not exiting

### ISSUE THREAD

**Issue by ramya-rao-a** [Azure/azure-service-bus-node#145]
**Describe the bug**
Completing a message from a streaming receiver with autoComplete enabled results in program not exiting

**To Reproduce**

- Use a streaming receiver (by default autoComplete will be enabled)
- Complete the message in the `onMessage` callback (`await message.complete()`)

This will result in the program not exiting. 

**Expected behavior**

Program exits 

**Additional context**

This is most likely due to the `autoComplete` feature trying to complete an already completed message

**Comment by HarshaNalluru** [comment]
For me, it took more time to exit(about 10-15 seconds more) when added `await message.complete()` call. 
But I didn't have to manually exit in either case(with and without `await message.complete()`).
Did you have to manually kill/close the execution ?

**Comment by HarshaNalluru** [comment]
**Bug**
   - It takes 20 seconds more time for the program to exit when we complete the message in the `onMessage` callback(`await message.complete()`) and if the autoComplete is true(by default autoComplete will be enabled).

**Reason for the Bug(Summary)**
   - When autoComplete is true(by default autoComplete will be enabled), `complete()` gets called for the second time and [timer attribute of _deliveryDispositionMap](https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L781)(`settimeout()`) gets overwritten by another `settimeout()`. 
   - The initial `settimeout()` is not cleared but overwritten. This causes the program to wait for another 20 seconds(`messageDispositionTimeout = 20000`) and then exits.
---------------------------------
**Detailed Explanation**
- [complete()](https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/serviceBusMessage.ts#L851) method in the serviceBusMessage.ts gets called twice for each message. 
   - First time, we call it from the sample as `brokeredMessage.complete()`.
   - And if autoComplete is true(which is by default), [complete()](https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L479
) is called for the second time from messageReceiver.ts.
   - The above `complete()` method, internally calls the [settleMessage()]( https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L769) in messageReceiver.ts.
       - Inside the `settleMessage()`, a `settimeout()` is assigned to the [timer attribute](https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L781) of [_deliveryDispositionMap](https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L792) of the service bus message.
- `complete()` is gets called twice and it calls `settleMessage()` once. So, `settleMessage()` gets called twice.
   - The `settimeout()` is assigned(to the `timer` attribute of `_deliveryDispositionMap`) when the `settleMessage()` method is called(by the `brokeredMessage.complete()`).
   - When autoComplete is true, `complete()` gets called for the second time and [timer attribute of _deliveryDispositionMap](https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L781)(`settimeout()`) gets overwritten by another `settimeout()`(when `complete()` calls `settleMessage()` for the second time.)
   - The second `settimeout()` gets [cleared]( https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L300) in the `_onSettled` event of messageReceiver.ts.
   - The first `settimeout()` got overwritten, but we did not explicitly call `clearTimeout()` to clear it.
   - This causes the program to wait for another 20 seconds(`messageDispositionTimeout = 20000`) and then exits.

---------------------------------

**Solution**
- Clear the existing `settimeout()` and then assign a new one.
- Or Keep a flag in the `_deliveryDispositionMap` which signifies the presence of `settimeout()`. 
     - If it is not present, we can set the `timer`. 
     - Otherwise, just don't overwrite the existing one.

---------------------------------



@ramya-rao-a @AlexGhiondea @ShivangiReja @ramya0820

**Comment by ramya-rao-a** [comment]
Wont https://github.com/Azure/azure-service-bus-node/blob/99b6bff5566f3f8499c8f38a97c7e9f37471cc55/lib/core/messageReceiver.ts#L300 in the `_onSettled` event listener clear the timeout for the first complete() call?

**Comment by HarshaNalluru** [comment]
Nope. Actually, the first timeout becomes an orphan and we have no way to access it.
The timer attribute is updated much before the execution of  `clearTimeout()`( in the `_onSettled`).
`clearTimeout()` only clears the `settimeout()` that is present as promise.timer, which is the overwritten `settimeout()`.

**Comment by HarshaNalluru** [comment]
`_onSettled()` is not called for the first `complete()` call.
The code in the `_onSettled()` is executed only once (_including the `clearTimeout()`_), that too after the second `complete()` call.

**Comment by ramya-rao-a** [comment]
I added some console.logs inside `_onSettled` and the place from where we autoComplete. Below is the flow of code I see

- `_onSettled()` called for the first complete() call made by the user
- Auto-complete code calls `complete()`
- (this takes some 20 seconds)
- Auto-complete code for `complete()` finishes without errors

Can you try with console.logs and see what you get? Be careful when debugging using breakpoints. Since the SDK uses timers in many places, the order that you see may not be the real order

**Comment by HarshaNalluru** [comment]
**The reason for above output**
- `_onSettled()` was executed after the two complete calls beacuse I've removed the `await` corresponding to the second `complete()` statement.

*Now, I've created a new branch from the master and made following observations.*

--------------------------------------


**Observations**
- The actual workflow is - `complete() -> _onSettled() -> second complete()`.
- We don't have to call the complete() method for the second time if it gets settled by the first call.
- I've observed that `delivery.remote_settled` gets updated to `true` when the message is settled.
- So, we can actually check the property( `delivery.remote_settled`) whether it is true or not just before making the second `complete()` call[that came from the autoComplete being true by default].

**Fix for the bug**
- `delivery.remote_settled` gets updated to `true` when the message gets settled.
- If `autoComplete` is `true`, `await bMessage.complete()` gets executed all the time without checking the  `delivery.remote_settled`. 
- We'd have to check the `delivery.remote_settled` and if it is false, then execute `await bMessage.complete()` 






The second `complete()` call.
https://github.com/Azure/azure-service-bus-node/blob/18f114dbfa0395b94de7380d8af9ceac26e8b560/lib/core/messageReceiver.ts#L479 
So, the above line of code should change to
```typescript 
if (!bMessage.delivery.remote_settled) {
       await bMessage.complete();
}
```
This would prevent calling `complete()` on a settled message.

**Comment by ramya-rao-a** [comment]
That makes perfect sense, can you submit a PR to fix this bug?

**Comment by ramya-rao-a** [comment]
Fixed with #154

**Comment by HarshaNalluru** [comment]
New PR for the same bug(at a different location) #157

**Comment by HarshaNalluru** [comment]
Fixed with https://github.com/Azure/azure-service-bus-node/pull/157
